compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -3,137 +3,230 @@
+  DiagnosticRelatedInformation(..),
+  List(..),
+  StoreItem(..),
+  Uri(..),
   noLocation,
   noRange,
   ideErrorText,
   ideErrorPretty,
   errorDiag,
   ideTryIOException,
   prettyFileDiagnostics,
-  prettyDiagnostic
+  prettyDiagnostic,
+  prettyDiagnosticStore,
+  defDiagnostic,
+  addDiagnostics,
+  filterSeriousErrors,
+  dLocation,
+  dFilePath,
+  filePathToUri,
+  getDiagnosticsFromStore
   ) where
 
-import Control.DeepSeq
 import Control.Exception
-import Data.Aeson (FromJSON, ToJSON)
+import Control.Lens (Lens', lens, set, view)
 import Data.Either.Combinators
-import Data.List.Extra
+import Data.Maybe as Maybe
+import Data.Foldable
+import qualified Data.Map as Map
 import qualified Data.Text as T
 import Data.Text.Prettyprint.Doc.Syntax
-import GHC.Generics
-import qualified Network.URI.Encode
+import Data.String (IsString(..))
 import qualified Text.PrettyPrint.Annotated.HughesPJClass as Pretty
+import           Language.Haskell.LSP.Types as LSP (
+    DiagnosticSeverity(..)
+  , Diagnostic(..)
+  , filePathToUri
+  , uriToFilePath
+  , List(..)
+  , DiagnosticRelatedInformation(..)
+  , Uri(..)
+  )
+import Language.Haskell.LSP.Diagnostics
 
 import Development.IDE.Types.Location
 
-ideErrorText :: FilePath -> T.Text -> Diagnostic
-ideErrorText absFile = errorDiag absFile "Ide Error"
-
-ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> Diagnostic
-ideErrorPretty absFile = ideErrorText absFile . T.pack . Pretty.prettyShow
-
-errorDiag :: FilePath -> T.Text -> T.Text -> Diagnostic
-errorDiag fp src msg =
-  Diagnostic
-  { dFilePath = fp
-  , dRange    = noRange
-  , dSeverity = Error
-  , dSource   = src
-  , dMessage  = msg
+ideErrorText :: FilePath -> T.Text -> LSP.Diagnostic
+ideErrorText fp = errorDiag fp "Ide Error"
+
+ideErrorPretty :: Pretty.Pretty e => FilePath -> e -> LSP.Diagnostic
+ideErrorPretty fp = ideErrorText fp . T.pack . Pretty.prettyShow
+
+errorDiag :: FilePath -> T.Text -> T.Text -> LSP.Diagnostic
+errorDiag fp src =
+  set dFilePath (Just fp) . diagnostic noRange LSP.DsError src
+
+-- | This is for compatibility with our old diagnostic type
+diagnostic :: Range
+           -> LSP.DiagnosticSeverity
+           -> T.Text -- ^ source
+           -> T.Text -- ^ message
+           -> LSP.Diagnostic
+diagnostic rng sev src msg
+    = LSP.Diagnostic {
+          _range = rng,
+          _severity = Just sev,
+          _code = Nothing,
+          _source = Just src,
+          _message = msg,
+          _relatedInformation = Nothing
+          }
+
+-- | Any optional field is instantiated to Nothing
+defDiagnostic ::
+  Range ->
+  T.Text -> -- ^ error message
+  LSP.Diagnostic
+defDiagnostic _range _message = LSP.Diagnostic {
+    _range
+  , _message
+  , _severity = Nothing
+  , _code = Nothing
+  , _source = Nothing
+  , _relatedInformation = Nothing
   }
 
-ideTryIOException :: FilePath -> IO a -> IO (Either Diagnostic a)
+-- | setLocation but with no range information
+dFilePath ::
```

Comment:
Do dFilePath and dLocation really need to be lenses? The setter seems useful. All the getters seem like they should work over a list rather than assuming its a single item anyway?

Reply:
I find it quite a bit easier to treat this as a maybe and the lens is a fairly nice way of doing that, it just means that it fits in better with our existing code. This should be fairly short lived

Reply:
As long as it's short lived I'm ok with it.